File: Novomber 2022/Day 26/C++ Code/Construct Binary Tree from String with bracket representation.cpp
//{ Driver Code Starts
//Initial Template for C++

#include <bits/stdc++.h>
using namespace std;

/* A binary tree node has data, pointer to left
child and a pointer to right child */
class Node
{
public:
    int data;
    Node *left, *right;
    Node(int data)
    {
        this->data = data;
        this->left = this->right = NULL;
    }
};

/* This function is to print the inorder of the tree  */
void inorder(Node *root)
{
    if (root == NULL)
        return;
    inorder(root->left);
    printf("%d ", root->data);
    inorder(root->right);
}



// } Driver Code Ends
//User function Template for C++

/*
class Node
{
public:
    int data;
    Node *left, *right;
    Node(int data)
    {
        this->data = data;
        this->left = this->right = NULL;
    }
};
*/


class Solution{
public:
    // function to construct tree from string
    Node *treeFromString(string str){
        // code here
        return f(str,0,str.length()-1);
    }
    
    Node* f(string s,int lft,int rt) {
        // System.out.println(s+" "+lft+" "+rt);
        if(lft>rt) return NULL;
        
        int i=lft;
        while(i<=rt) {
            char c = s[i];
            if(c>='0' && c<='9') {
                i++;
            } else {
                break;
            }
        }
        
        Node* ans = new Node(stoi(s.substr(lft,i-lft)));
        
        int j=i;
        stack<char> stk;
        while(j<=rt) {
            char c = s[j];
            if(c==')') {
                stk.pop();
            } else if(c=='(') {
                stk.push('(');
            }
            if(stk.size()==0) break;
            j++;
        }
        
        ans->left = f(s,i+1,j-1);
        if(j+2<rt) ans->right = f(s,j+2,rt-1);
        return ans;
    }
    
};

//{ Driver Code Starts.

// Driver Code
int main()
{
    int t;
    cin >> t;
    while (t--)
    {
        string str;
        cin >> str;
        Solution obj;
        Node *root = obj.treeFromString(str);
        inorder(root);
        cout << "\n";
    }
    return 0;
}

// } Driver Code Ends

File: Novomber 2022/Day 26/JAVA Code/Construct Binary Tree from String with bracket representation.java
//{ Driver Code Starts
import java.io.*;
import java.util.*;


class Node
{
    int data;
    Node left;
    Node right;
    
    Node(int data)
    {
        this.data = data;
        left = null;
        right = null;
    }
    
    public static Node buildTree(String str)
    {
        // Corner Case
        if(str.length()==0 || str.charAt(0)=='N')
            return null;
        
        // Creating array of Strings from input 
        // String after spliting by space
        String ip[] = str.split(" ");
        
        // Create the root of the tree
        Node root = new Node(Integer.parseInt(ip[0]));
        
        // Push the root to the queue
        Queue<Node> queue = new LinkedList<>(); 
        queue.add(root);
        
        // Starting from the second element
        int i = 1;
        while(queue.size()>0 && i < ip.length)
        {
            
            // Get and remove the front of the queue
            Node currNode = queue.peek();
            queue.remove();
                
            // Get the current node's value from the string
            String currVal = ip[i];
                
            // If the left child is not null
            if(!currVal.equals("N"))
            {
                    
                // Create the left child for the current node
                currNode.left = new Node(Integer.parseInt(currVal));
                // Push it to the queue
                queue.add(currNode.left);
            }
                
            // For the right child
            i++;
            if(i >= ip.length)
                break;
                
            currVal = ip[i];
                
            // If the right child is not null
            if(!currVal.equals("N"))
            {
                    
                // Create the right child for the current node
                currNode.right = new Node(Integer.parseInt(currVal));
                    
                // Push it to the queue
                queue.add(currNode.right);
            }
            i++;
        }
        
        return root;
    }
    
    public static Node inputTree(BufferedReader br) throws IOException
    {
        return buildTree(br.readLine().trim());
    }
    
    public static void inorder(Node root)
    {
        if (root == null)
           return;
        inorder(root.left);
        System.out.print(root.data + " ");
        inorder(root.right);
    }
}

class GFG {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t;
        t = Integer.parseInt(br.readLine());
        while(t-- > 0){
            
            String s; 
            s = br.readLine();
            
            Solution obj = new Solution();
            Node res = obj.treeFromString(s);
            
            Node.inorder(res);
            System.out.println();
            
        }
    }
}

// } Driver Code Ends


/*

Definition for Binary Tree Node
class Node
{
    int data;
    Node left;
    Node right;
    
    Node(int data)
    {
        this.data = data;
        left = null;
        right = null;
    }
}
*/

class Solution {
    public static Node treeFromString(String s) {
        // code here
        return f(s,0,s.length()-1);
    }
    
    public static Node f(String s,int lft,int rt) {
        // System.out.println(s+" "+lft+" "+rt);
        if(lft>rt) return null;
        
        int i=lft;
        while(i<=rt) {
            char c = s.charAt(i);
            if(c>='0' && c<='9') {
                i++;
            } else {
                break;
            }
        }
        
        Node ans = new Node(Integer.parseInt(s.substring(lft,i)));
        
        int j=i;
        Stack<Character> stack = new Stack<>();
        while(j<=rt) {
            char c = s.charAt(j);
            if(c==')') {
                stack.pop();
            } else if(c=='(') {
                stack.push('(');
            }
            if(stack.size()==0) break;
            j++;
        }
        
        ans.left = f(s,i+1,j-1);
        if(j+2<rt) ans.right = f(s,j+2,rt-1);
        return ans;
    }
}
        
